README
= Description
   A Ruby interface for getting cpu information.

= Installation
  gem install sys-cpu

= Notes
== Solaris
  Currently there is no 'processors()' iterative method for multi-cpu systems.
  I plan to add it this in a future release.
   
== OS X
  The CPU.model method returns very limited information. I do not yet know
  how to get more detailed information.

== Linux
  This is pure Ruby. This version reads information out of /proc/cpuinfo and
  /proc/loadavg, so if /proc isn't mounted it won't work.

  The key-value information in /proc/cpuinfo is stored internally (i.e. in
  memory) as an array of hashes when you first 'require' this package.  This
  overhead is exceptionally minimal, given that your average cpuinfo file
  contains less than 1k of text (and I don't store whitespace or newlines).

  The text documentation for Linux is dynamically generated during the
  build process because the fields vary depending on your setup.  So, don't
  look at it until *after* you've installed it.  You will see a doc/linux.txt
  file after you run 'rake install' (via install.rb).

== HP-UX
  Unlike other platforms, you can get load averages for an individual cpu in
  multi-cpu systems. See documentation for more details.

  Note that version 0.7.x and later will not work on HP-UX because of the
  switch to FFI and the lack of a testing platform. However, version 0.6.x
  will work just fine.

== MS Windows
  This is a pure Ruby implementation using the win32ole package + WMI. The C
  version has been scrapped. 

  As of version 0.5.0, the CPU.usage method has been removed in favor of the
  CPU.load_avg method.  This does not (currently) use a perf counter, so there
  is no longer any delay.  Also, the 'processors' method has been
  added and the 'supported' method has been dropped.  See the documentation
  for other changes.
   
= Acknowledgements
  Thanks go to the MPlayer team for some source code that helped me on
  certain versions of FreeBSD in the original C version.

= Known Bugs
  None that I'm aware of. Please report bugs on the project page at
  https://github.com/djberg96/sys-cpu

= Future Plans
  Add iterative CPU.processors method.
  Add more information in general, such as what 'prtdiag' shows.

= License
  Artistic 2.0

= Copyright
  (C) 2003-2015 Daniel J. Berger, All Rights Reserved

= Warranty
  This package is provided "as is" and without any express or
  implied warranties, including, without limitation, the implied
  warranties of merchantability and fitness for a particular purpose.

= Author
  Daniel J. Berger


CHANGES
== 0.8.1 - 4-Nov-2018
* Added metadata to the gemspec.
* Fixed missing hyphen in license name.

== 0.8.0 - 17-Oct-2018
* Switched license to Apache 2.0.
* Updated documentation for the cpu_stats method on Linux.
* The VERSION constant is now frozen.
* Now uses require_relative internally where needed.

== 0.7.2 - 5-Sep-2015
* Replaced a global array with a constant in the Linux version. Thanks go
  to Yorick Peterse for the patch.
* Added method comments back to the Unix version.
* Added a sys-cpu.rb stub file for your convenience.
* This gem is now signed.

== 0.7.1 - 2-May-2013
* Added a workaround for a win32ole bug in the Windows version.
* Reorganized code so that there is now a single gem rather than three
  separate platform gems.
* Updated test-unit dependency which let me simplify the test files a bit.

== 0.7.0 - 14-Dec-2011
* Code base now uses FFI. However, HP-UX is not currently supported.
  HP-UX users should continue to use the 0.6.x version. Note that
  the code base remains unchanged on Linux and Windows.
* The cpu_type method has been replaced with the architecture method
  on Solaris to keep the interface more in line with other platforms.
* The type method has been changed to cpu_type on Windows.
* Some Rakefile and test suite updates.

== 0.6.4 - 27-Sep-2011
* The CPU.freq method now works on OSX.
* The CPU.model method on OSX has been altered. Previously it
  returned the machine model. However, the information is limited.
* Fixed a couple unused variable warnings for the BSD/OSX code.
* The Linux and Windows gems now have a 'universal' architecture.
* Refactored the clean task in the Rakefile.

== 0.6.3 - 9-Oct-2010
* Fixed a bug in the install.rb file and refactored it a bit. Thanks go
  to Di-an Jan for the spot. Note, however, that this file will eventually
  be removed and/or integrated into the Linux source file.
* Fixed the example Rake task, and refactored some of the other tasks.
* Fixed and updated the CPU.architecture method on MS Windows to handle IA64
  and x64 architectures.

== 0.6.2 - 1-Jan-2010
* Fixed a bug in the cpu_freq function (which would only be noticed on
  certain platforms in cases where the CPU.freq method failed). Thanks
  go to Edho P Arief for the spot.
* Explicitly add sys/param.h on OpenBSD, and default to the HW_MODEL mib
  because HW_MACHINE does not exist. Thanks go to Edho P Arief for the patch.
* Updated my support notes. In short I will support 1.8.6 and 1.9.x. I will
  not support any 1.8.x branch later than 1.8.6.
* Removed redundant information in the various .txt files that is already
  covered in the README file.
* Test files renamed.
* Added test-unit 2.x as a development dependency.

== 0.6.1 - 4-Jan-2009
* Fix for OS X 10.5.x. Thanks go to Victor Costan for the spot and the patch.
* Updated the gemspec and some other minor changes.
* On MS Windows the impersonation level is now explicitly set to 'impersonate'
  to avoid issues where systems might be using an older version of WMI.

== 0.6.0 - 26-Apr-2007
* Added support for most BSD flavors, including OS X. The freebsd.c file is
  now just bsd.c.
* The CPU.type method for Solaris has been changed to CPU.cpu_type to avoid
  conflicting with the Object.type method.
* Added a Rakefile. There are now tasks for building, testing and installing,
  among other things. Run 'rake -T' to check your options.
* Many internal directory layout changes - C source files are now under the
  'ext' directory.
* Improved RDoc comments in the C source files.
* Changed CPUError to CPU::Error.

== 0.5.5 - 17-Nov-2006
* Fixed a bug in the Linux version where there could be a key but no
   associated value, causing a String#strip! call to fail.  Now the value is
   simply left at nil.
* Refactored the CPU.processors method on Linux, including the removal of '?'
  from the CPUStruct members for the boolean methods (Ruby doesn't like them).
* Minor tweaks and updates to the documentation, including the README.
* Modified the gemspec so that it sets the platform properly for Linux
  and Windows.

== 0.5.4 - 12-Jul-2006
* Added a gemspec (and a gem on RubyForge).
* The CPU.architecture method on HP-UX now returns nil if it cannot be
  determined instead of "Unknown" for consistency with the other
  platforms.
* Inlined the RDoc and made some minor cosmetic source code changes.

== 0.5.3 - 4-May-2006
* Fixed in a bug in the Solaris version where CPU.load_avg returned bad values
  when compiled in 64 bit mode.  Thanks go to James Hranicky for the spot and
  both James Hranicky and Peter Tribble (via comp.unix.solaris) for patches.
* Made some modifications to the test suite.  You know longer need to know
  which test suite to run.  All platforms now use 'ts_all.rb', which will run
  the appropriate test case behind the scenes.

== 0.5.2 - 24-Jun-2005
* Bug fixed on Linux where empty lines could cause problems.  This affected
  both the install.rb and linux.rb files, though it only seems to have been
  an issue on the 2.6+ kernel.
* Altered the behavior of the CPU.cpu_stats method on Linux.  Now, only the
  numbered cpu entries return associated data, unless the numberless entry
  is the only entry.
* Added a sample program for Linux under the 'examples' directory.

== 0.5.1 - 5-May-2005
* Fixed a potential bug in the FreeBSD version of CPU.model.
* Eliminated some warnings in the FreeBSD version.
* Moved examples directory to the toplevel package directory.
* Renamed and updated the sample scripts.
* Added a sample script for FreeBSD.
* Removed the INSTALL file.  That information is now included in the README.
* Made the CHANGES, README, and .txt files rdoc friendly.
* The dynamic documentation generation for Linux has been altered.  Anything
  relating to rd2 has been removed.  The doc file generated is now
  doc/linux.txt.
* Some $LOAD_PATH setup changes in the unit tests.

== 0.5.0 - 26-Jun-2004
* Now requires Ruby 1.8.0 or later.
* FreeBSD support added.
* Replaced 'CPUException' with 'CPUError'

ext/bsd/bsd.c
/*****************************************************************************
 * bsd.c (cpu.c) - sys-cpu extension for the various BSD flavors and OS X.
 *
 * Author: Daniel J. Berger
 *
 * Interface to provide various types of cpu information.
 * Based on the Unix::Processors Perl module (Wilson Snyder) with ideas from
 * Sys::CPU (Matt Sanford) and Solaris::Kstat (Alan Burlison) as well.
 * OS X 10.5+ patch for uptime by Victor Costan.
 *
 * Portions of this code lifted from the MPlayer source (cpuinfo.c).
 *****************************************************************************/
#include <ruby.h>
#include "version.h"

#ifdef HAVE_KVM_H
#include <kvm.h>
#endif

#if defined (__OpenBSD__)
#include <sys/param.h>
#endif

#include <sys/sysctl.h>
#include <sys/types.h>
#include <string.h>
#include <errno.h>

#ifndef MISSING_USLEEP
#include <unistd.h>
#endif

#if defined(__MACH__) && defined(__APPLE__)
#include <mach/machine.h>
#endif

VALUE cCPUError;

/****************************************************************************
 * Used for FreeBSD 4.x to determine CPU clock speed. Borrowed from cpuinfo.c
 * in the MPlayer source code.
 ****************************************************************************/
#if defined (__FreeBSD__) && (__FreeBSD__ < 5 )
static int64_t rdtsc(void){
   unsigned int i, j;
#define RDTSC  ".byte 0x0f, 0x31; "
   asm(RDTSC : "=a"(i), "=d"(j) : );
   return ((int64_t)j<<32) + (int64_t)i;
}
#endif

/*
 * call-seq:
 *    CPU.load_average
 *
 * Returns an array of three floats indicating the 1, 5 and 15 minute load
 * average.
 */
static VALUE cpu_load_avg(VALUE klass){
   int n;
   VALUE v_num_array = rb_ary_new();

#ifdef HAVE_KVM_H
  int max = 3;
  kvm_t* k;
  double avgs[3];

  k = malloc(sizeof(kvm_t*));

  if(!kvm_getloadavg(k, avgs, max)){
    free(k);
    rb_raise(cCPUError, "error calling kvm_getloadavg(): %s", strerror(errno));
  }

  for(n = 0; n < 3; n++)
    rb_ary_push(v_num_array, rb_float_new(avgs[n]));

  free(k);
#else
	struct loadavg k;
	size_t len = sizeof(k);

#ifdef HAVE_SYSCTLBYNAME
  if(sysctlbyname("vm.loadavg", &k, &len, NULL, 0))
    rb_raise(cCPUError, "error calling sysctlbyname(): %s", strerror(errno));
#else
  int mib[2];
  mib[0] = CTL_HW;
  mib[1] = VM_LOADAVG;

  if(sysctl(mib, 2, &k, &len, NULL, 0))
    rb_raise(cCPUError, "error calling sysctl(): %s", strerror(errno));
#endif
  for(n = 0; n < 3; n++)
    rb_ary_push(v_num_array, rb_float_new(k.ldavg[n] / (float)k.fscale));
#endif

  return v_num_array;
}

/*
 * call-seq:
 *    CPU.num_cpu
 *
 * Returns the number of cpu's on your system. Note that each core on
 * multi-core systems are counted as a cpu, e.g. one dual core cpu would
 * return 2, not 1.
 */
static VALUE cpu_num(VALUE klass){
   int num_cpu;
   size_t len = sizeof(num_cpu);

#ifdef HAVE_SYSCTLBYNAME
   if(sysctlbyname("hw.ncpu", &num_cpu, &len, NULL, 0))
      rb_raise(cCPUError, "error calling sysctlbyname(): %s", strerror(errno));
#else
   int mib[2];
   mib[0] = CTL_HW;
   mib[1] = HW_NCPU;

   if(sysctl(mib, 2, &num_cpu, &len, NULL, 0))
      rb_raise(cCPUError, "error calling sysctl(): %s", strerror(errno));
#endif

   return INT2NUM(num_cpu);
}

/*
 * call-seq:
 *    CPU.model
 *
 * Returns a string indicating the cpu model.
 *--
 * On OSX I use the hw.cputype instead of hw.model because OSX returns the
 * machine model instead of the cpu model for some reason. Initial attempts
 * to add the cpusubtype as well were unreliable.
 */
static VALUE cpu_model(VALUE klass){
  char model[64];
  size_t len = sizeof(model);

#ifdef HAVE_SYSCTLBYNAME
#if defined(__MACH__) && defined(__APPLE__)
  int cpu_type;
  len = sizeof(cpu_type);

  if(sysctlbyname("hw.cputype", &cpu_type, &len, NULL, 0))
#else
  if(sysctlbyname("hw.model", &model, &len, NULL, 0))
#endif
    rb_raise(cCPUError, "error calling sysctlbyname(): %s", strerror(errno));
#else
  int mib[2];

  mib[0] = CTL_HW;
  mib[1] = HW_MODEL;

  if(sysctl(mib, 2, &model, &len, NULL, 0))
    rb_raise(cCPUError, "error calling sysctl(): %s", strerror(errno));
#endif

#if defined(__MACH__) && defined(__APPLE__)
  // Intel and PowerPC only.
  switch(cpu_type){
    case CPU_TYPE_X86:
    case CPU_TYPE_X86_64:
      strcpy(model, "Intel");
      break;
    case CPU_TYPE_POWERPC:
    case CPU_TYPE_POWERPC64:
      strcpy(model, "PowerPC");
      break;
    default:
      strcpy(model, "Unknown"); 
  }

  // TODO: Add the subtype.
#endif

  return rb_str_new2(model);
}

/*
 * call-seq:
 *    CPU.architecture
 *
 * Returns the cpu's architecture. On most systems this will be identical
 * to the CPU.machine method. On OpenBSD it will be identical to the CPU.model
 * method.
 */
static VALUE cpu_architecture(VALUE klass){
   char arch[32];
   size_t len = sizeof(arch);

#ifdef HAVE_SYSCTLBYNAME
#if defined(__MACH__) && defined(__APPLE__)
   if(sysctlbyname("hw.machine", &arch, &len, NULL, 0))
      rb_raise(cCPUError, "error calling sysctlbyname(): %s", strerror(errno));
#else
   if(sysctlbyname("hw.machine_arch", &arch, &len, NULL, 0))
      rb_raise(cCPUError, "error calling sysctlbyname(): %s", strerror(errno));
#endif
#else
   int mib[2];
   mib[0] = CTL_VM;
#ifdef HW_MACHINE_ARCH
   mib[1] = HW_MACHINE_ARCH;
#else
   mib[1] = HW_MODEL;
#endif

   if(sysctl(mib, 2, &arch, &len, NULL, 0))
      rb_raise(cCPUError, "error calling sysctl(): %s", strerror(errno));
#endif

   return rb_str_new2(arch);
}

/*
 * call-seq:
 *    CPU.machine
 *
 * Returns the cpu's class type. On most systems this will be identical
 * to the CPU.architecture method. On OpenBSD it will be identical to the
 * CPU.model method.
 */
static VALUE cpu_machine(VALUE klass){
   char machine[32];
   size_t len = sizeof(machine);

#ifdef HAVE_SYSCTLBYNAME
   if(sysctlbyname("hw.machine", &machine, &len, NULL, 0))
      rb_raise(cCPUError, "error calling sysctlbyname(): %s", strerror(errno));
#else
   int mib[2];
   mib[0] = CTL_HW;
#ifdef HW_MACHINE_ARCH
   mib[1] = HW_MACHINE;
#else
   mib[1] = HW_MODEL;
#endif

   if(sysctl(mib, 2, &machine, &len, NULL, 0))
      rb_raise(cCPUError, "error calling sysctl(): %s", strerror(errno));
#endif

   return rb_str_new2(machine);
}

/*
 * call-seq:
 *    CPU.freq
 *
 * Returns an integer indicating the speed (i.e. frequency in Mhz) of the cpu.
 */
static VALUE cpu_freq(VALUE klass){
   int mhz;
#if defined (__FreeBSD__) && (__FreeBSD__ < 5)
   int64_t tsc_start, tsc_end;
   struct timeval tv_start, tv_end;
   int usec_delay;

   tsc_start = rdtsc();
   gettimeofday(&tv_start,NULL);
#ifdef MISSING_USLEEP
   sleep(1);
#else
   usleep(100000);
#endif
   tsc_end = rdtsc();
   gettimeofday(&tv_end,NULL);

   usec_delay = 1000000 * (tv_end.tv_sec - tv_start.tv_sec)
      + (tv_end.tv_usec - tv_start.tv_usec);

   mhz = ((tsc_end - tsc_start) / usec_delay);
#else
   size_t len = sizeof(mhz);
#ifdef HAVE_SYSCTLBYNAME
#if defined(__MACH__) && defined(__APPLE__)
   size_t freq;
   len = sizeof(freq);

   if(sysctlbyname("hw.cpufrequency", &freq, &len, 0, 0))
      rb_raise(cCPUError, "error calling sysctlbyname(): %s", strerror(errno));

   mhz = freq / 1000000;
#else
   if(sysctlbyname("hw.clockrate", &mhz, &len, 0, 0))
      rb_raise(cCPUError, "error calling sysctlbyname(): %s", strerror(errno));
#endif
#else
   int mib[2];

   mib[0] = CTL_KERN;
   mib[1] = KERN_CLOCKRATE;

   if(sysctl(mib, 2, &mhz, &len, NULL, 0))
      rb_raise(cCPUError,"error calling sysctlbyname(): %s", strerror(errno));
#endif
#endif

   return INT2NUM(mhz);
}

void Init_cpu()
{
   VALUE mSys, cCPU;

   /* The Sys module serves as a toplevel namespace only */
   mSys = rb_define_module("Sys");

   /* The CPU class provides class methods for obtaining CPU information */
   cCPU = rb_define_class_under(mSys, "CPU", rb_cObject);

   /* The CPU::Error Exception class is raised whenever any of the CPU class
    * methods fail.
    */
   cCPUError = rb_define_class_under(cCPU, "Error", rb_eStandardError);

   /* 0.6.4: The version of the sys-cpu library */
   rb_define_const(cCPU, "VERSION", rb_str_new2(SYS_CPU_VERSION));

   /* Class Methods */
   rb_define_singleton_method(cCPU, "architecture", cpu_architecture, 0);
   rb_define_singleton_method(cCPU, "freq", cpu_freq, 0);
   rb_define_singleton_method(cCPU, "load_avg", cpu_load_avg, 0);
   rb_define_singleton_method(cCPU, "machine", cpu_machine, 0);
   rb_define_singleton_method(cCPU, "model", cpu_model, 0);
   rb_define_singleton_method(cCPU, "num_cpu", cpu_num, 0);
}
